---
title: "Training and predicting on multiple data sets"
author: "Michelle N. Ngo"
date: 'Last compiled: `r format(Sys.Date(), "%B %d, %Y")`'
header-includes:
    - \usepackage {hyperref}
    - \hypersetup {colorlinks = true, linkcolor = purple, urlcolor = pink}
geometry: "left=1cm,right=1cm,top=1cm,bottom=1cm"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Training and predicting on multiple data sets}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include=FALSE, echo=FALSE}
library(knitr)
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Set-up

Load the package.

```{r setup}
library(tauFisher)
library(dplyr)
set.seed(123)
```

# Import data

`tauFisher` requires:

  1. A matrix of gene expression data with rows as genes and columns as samples.  
  2. A vector of the circadian time for each sample. 
  

For this tutorial, we will train `tauFisher` on the circadian time of suprachiasmatic nucleus (SCN) transcriptomic samples from C57BL/6J mice. These mice were fed normal chow through an entire circadian cycle (24 hours) and samples were collected every four hours. A total of six time points with three replicates each (18 samples) were sequenced on the Illumina HiSeq 4000. 


For more details, see [Tognini P *et al.* (2020)](https://doi.org/10.1073/pnas.2016589117). 
The data set can be downloaded from the GEO database: [accession number GSE157077](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157077).

```{r load_bulk}
bulk_file <- system.file("extdata", "GSE157077_mouse_scn_control.tsv", 
                    package = "tauFisher", mustWork = TRUE)

bulk_data <- utils::read.delim(file = bulk_file, stringsAsFactors = FALSE)
print(dim(bulk_data))

print(bulk_data[1:5, 1:5])
```


We will use the trained model to predict the circadian time of SCN transcriptomic samples from wild-type C57BL/6J mice. Samples were collected every four hours over two days for a total of twelve time points. These were sequenced on the HiSeq X Ten (10x Chromium).


For more details, see [Wen S *et al.* (2020)](https://doi.org/10.1038/s41593-020-0586-x). 
The single cell data set can be downloaded from the GEO database: [accession number GSE132608](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE132608).


```{r load_pseudobulk}
pseudobulk_file <- system.file("extdata", "GSE132608_scn_pseudobulk.RData", 
                    package = "tauFisher", mustWork = TRUE)

load(pseudobulk_file)
print(dim(pseudobulk_data))

print(pseudobulk_data[1:5, 1:5])
```

# Pre-process data 

All genes in the data matrix should be unique; if they are not, we will take their average expression. 

```{r preprocess}
# Take the average expression if there are non-unique genes
bulk <- bulk_data[stats::complete.cases(bulk_data[,-1]), ] %>%
  dplyr::mutate(ID = toupper(ID)) %>%
  dplyr::group_by(ID) %>%
  dplyr::summarise_all(mean)  %>%
  as.data.frame()

# Set the rownames to be the genes and remove the ID column 
rownames(bulk) <- bulk$ID
bulk <- bulk[,-1]
```

For this data set, the column names are the sample times and replicate indicator. 
We will parse the column names to obtain a vector of time points. 

```{r preprocess_time}
time <- as.numeric(vapply(stringr::str_split(colnames(bulk), "_"),
                          '[', 2, FUN.VALUE = character(1)) )
replicate <- as.numeric(vapply(stringr::str_split(colnames(bulk), "_"),
                          '[', 4, FUN.VALUE = character(1)) )
```

```{r adjust_time}
# adjust time - each replicate is now the next 'period' so there's 3 sets
time_adj <- time + 24*(as.numeric(replicate) - 1)

bulk_adj = bulk
colnames(bulk_adj) <- time_adj
bulk_adj = bulk_adj[, order(as.numeric(colnames(bulk_adj)))]
time_adj <- time_adj[order(as.numeric(time_adj))]
```


Note that the pseudobulk data set that we will use for prediction has already been processed. For raw single cell data, you will want to sum the expression levels over all the cells for each time point to obtain a pseudobulk data matrix. 

Here, we just need to parse the column names to obtain the vector of time points for the pseudobulk data. 

```{r preprocess_pseudo_time}
pseudobulk_data[is.na(pseudobulk_data)] <- 0
pseudobulk_time <- as.numeric(vapply(stringr::str_split(colnames(pseudobulk_data), "CT"),
                                     '[', 2, FUN.VALUE = character(1))) 

# View test inputs
print(pseudobulk_data[1:5, 1:5])
print(pseudobulk_time)
```


# `tauFisher`

We will train `tauFisher` on the bulk data set and test on the single cell.   

## Find the periodic genes 

For the two data sets we are working with, we do not have any missing data points. 

Note that if there are missing time points, `JTK_Cycle` requires the data set to have a column of `NA`s for each missing point. 
The columns are also assumed to be in order of time, for example: `CT0_REP1, CT0_REP2, CT4_REP1, CT4_REP2, ... `.


For more details, please see the documentation for the `meta2d` function in `MetaCycle` or the `JTK_Cycle` package. 


We will use both `Lomb-Scargle` and `JTK_Cycle` to find the periodic genes.
`meta2d` will produce a text file (location and name given by parameter `out_file`) and a folder (location and name given by parameter `out_dir`) with the results. 

```{r see_inputs}
# View the inputs
print(bulk_adj[1:5, 1:5])
print(time_adj)
```


```{r metacycle, eval=FALSE}
run_metacycle(df=bulk_adj, 
              timepoints=time_adj, 
              out_file="../inst/extdata/GSE157077_full_adj.txt", 
              out_dir="../inst/extdata/GSE157077_full_adj_meta2d", 
              method=c("JTK", "LS") )
```

```{r periodic_genes}
# Don't set a threshold for p-value
meta2d_file = "../inst/extdata/GSE157077_full_adj_meta2d/meta2d_GSE157077_full_adj.txt"
genes = find_periodic_genes(input_file=meta2d_file,
                            test_genes=rownames(pseudobulk_data),
                            per_method=c("LS", "JTK"),
                            thres=1)

print(genes)
```


## Train `tauFisher`

For simplicity, we will proceed with the periodic genes identified by JTK_Cycle.

```{r train}
trained_JTK <- train_tauFisher(periodic_genes=genes$JTK,
                               train_df=bulk_adj,
                               train_time=time_adj,
                               nrep=3,
                               numbasis=5)
```

The trained model will output a list with three items:

1. The genes used for circadian time prediction.

    ```{r}
    trained_JTK$Genes
    ```

2. The PCA embedding. 
3. The trained model.

## Predict circadian time 


```{r test, eval=T}
result_JTK <- test_tauFisher(chosen_genes=trained_JTK$Genes,
                             test_df=pseudobulk_data,
                             test_time=pseudobulk_time,
                             pca_embedding=trained_JTK$PCA_embedding,
                             trained_model=trained_JTK$Model)
```

If `tauFisher` is unable to predict the time of your samples due to the test data having 

## Evaluate predictions

The prediction is labeled correct if the predicted time is within 2 hours of the true time. 

```{r evaluate}
calc_error(result_JTK$truth, result_JTK$predicted)
```

